cdc/processor/agent.go

```diff
@@ -0,0 +1,369 @@
+			log.Warn("Processor Agent received error from message handler",
+				zap.Error(err))
+			return errors.Trace(err)
+		default:
+		}
+	}
+
+	if err := a.BaseAgent.Tick(ctx); err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+func (a *agentImpl) FinishTableOperation(
+	ctx context.Context,
+	tableID model.TableID,
+) (bool, error) {
+	done, err := a.trySendMessage(
+		ctx, a.ownerCaptureID,
+		model.DispatchTableResponseTopic(a.changeFeed),
+		&model.DispatchTableResponseMessage{ID: tableID})
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+	return done, nil
+}
+
+func (a *agentImpl) SyncTaskStatuses(
+	ctx context.Context,
+	running, adding, removing []model.TableID,
+) (bool, error) {
+	done, err := a.trySendMessage(
+		ctx,
+		a.ownerCaptureID,
+		model.SyncTopic(a.changeFeed),
+		&model.SyncMessage{
+			ProcessorVersion: version.ReleaseSemver(),
+			Running:          running,
+			Adding:           adding,
+			Removing:         removing,
+		})
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+	return done, nil
+}
+
+func (a *agentImpl) SendCheckpoint(
+	ctx context.Context,
+	checkpointTs model.Ts,
+	resolvedTs model.Ts,
+) (bool, error) {
+	done, err := a.trySendMessage(
+		ctx,
+		a.ownerCaptureID,
+		model.CheckpointTopic(a.changeFeed),
+		&model.CheckpointMessage{
+			CheckpointTs: checkpointTs,
+			ResolvedTs:   resolvedTs,
+		})
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+	return done, nil
+}
+
+// Barrier returns whether there is a pending message not yet acknowledged by the owner.
+// Please refer to the documentation on the ProcessorMessenger interface.
+func (a *agentImpl) Barrier(_ context.Context) (done bool) {
+	defer func() {
+		if done {
+			return
+		}
+
+		sinceLastAdvanced := a.clock.Since(a.barrierLastCleared)
+		if sinceLastAdvanced > barrierNotAdvancingWarnDuration && a.logRateLimiter.Allow() {
+			log.Warn("processor send barrier not advancing, report a bug if this log repeats",
+				zap.String("changefeed-id", a.changeFeed),
+				zap.String("owner-id", a.ownerCaptureID),
+				zap.Duration("duration", sinceLastAdvanced))
+		}
+	}()
+
+	if a.barrierLastCleared.IsZero() {
+		a.barrierLastCleared = a.clock.Now()
+	}
+
+	if a.ownerCaptureID == "" {
+		// We should wait for the first owner to contact us.
+		// We need to wait for the sync request anyways, and
+		// there would not be any table to replicate for now.
+		log.Debug("waiting for owner to request sync",
+			zap.String("changefeed-id", a.changeFeed))
+		return false
+	}
+
+	client := a.messageRouter.GetClient(a.ownerCaptureID)
+	if client == nil {
+		// Client not found for owner.
+		// Note that if the owner is eventually gone,
+		// OnOwnerChanged will reset the barriers.
+		return false
+	}
+	for topic, waitSeq := range a.barrierSeqs {
+		actualSeq, ok := client.CurrentAck(topic)
+		if !ok {
+			return false
+		}
+		if actualSeq >= waitSeq {
+			delete(a.barrierSeqs, topic)
+		} else {
+			return false
+		}
+	}
```

Comment:
Could it happen this scenario, given two topics, t-a, t-b, and sequence enqueue as follows
```
t-a wait 10, t-b wait 11, t-a wait 12, t-b wait 13...
             t-a ack 10 , t-b-ack 11, t-a ack 12  ...            
```
which leads to `Barrier` always return false.

Reply:
This scenario will not happen, because `Barrier` returning false will prevent sending the checkpoint, which will prevent the Owner from sending any further commands, so the processor will not reply anything.

I will think more about this scenario and try to refactor so that this kind of problem will not be introduced in the future.